### PR TITLE
Updated Program.py

### PR DESCRIPTION
**Pull request by agarwal-aryan**
Renamed `Program` to `NestedProgram`.

Issue #342 

- [x] I have read the CONTRIBUTING document.

##Summary
Updated `Program` to `NestedProgram` and related references like `get_program` to `get_nested_program` and many more.

### CODE REVIEW

**Comment by pacomf** [comment]
@agarwal-aryan you need to update tests files as well, to rename reference there too :) (See tests failing about it: https://github.com/Qiskit-Extensions/quantum-serverless/actions/runs/4503957936/jobs/7928024705?pr=347)

File: client/quantum_serverless/core/program.py [hunk 1]
```diff
@@ -13,18 +13,18 @@
 
 """
 =================================================
```
Comment (IceKhan13) [inline client/quantum_serverless/core/program.py:15]: Also keep in mind that formatting for rst files is strict, length of line should match length or `===` segment.

Comment (IceKhan13) [reply client/quantum_serverless/core/program.py:15]: and thank you for talking this issue @agarwal-aryan !

Comment (agarwal-aryan) [reply client/quantum_serverless/core/program.py:15]: @IceKhan13 can you please tell me what is to be done here as, I am not able to understand your point?

**Comment by agarwal-aryan** [comment]
@pacomf Do I have to change the name of the file also?

File: client/quantum_serverless/core/program.py (hunk 1 above)
Comment (Tansito) [reply client/quantum_serverless/core/program.py:15]: I don't know if this is a better explanation but what Iskandar means is that the lenght of the previous line should match with length of the `===` line. Example:

```rst
Quantum serverless nested program
===========================
```

Like both lines has the same length (it's just an example). But the linter should give you clues about these things, just run `tox -elint` and will tell you the problems to fix. Thank you @agarwal-aryan 😄

File: client/tests/core/test_nested_program_repository.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ def text(self):
     @property
     def ok(self):  # pylint: disable=invalid-name
         """Status of response."""
-        return True
+        return TrueProgram
```
Comment (Tansito) [inline client/tests/core/test_nested_program_repository.py:70]: I think must be `True` yet.

File: client/quantum_serverless/core/nested_program.py [hunk 3]
```diff
@@ -146,20 +146,20 @@ def __init__(
         self._host = host or os.environ.get(REPO_HOST_KEY, "http://localhost")
         self._port = port or os.environ.get(REPO_PORT_KEY, 80)
         self._token = token
-        self._base_url = f"{self._host}:{self._port}/v1/api/nested-programs/"
+        self._base_url = f"{self._host}:{self._port}/v1/api/nested-nested_programs/"
```
Comment (Tansito) [inline client/quantum_serverless/core/nested_program.py:149]:
```suggestion
        self._base_url = f"{self._host}:{self._port}/v1/api/nested-programs/"
```

File: client/tests/core/test_nested_program_repository.py [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
 import shutil
 from pathlib import Path
 from unittest import TestCase, mock
 
-from quantum_serverless.core.program import ProgramRepository, Program
+from quantum_serverless.core.nested_program import NestedProgramRepository, NestedProgram
 
 responses = {
-    "http://localhost:80/v1/api/nested-programs/": {
+    "http://localhost:80/v1/api/nested-nested_programs/": {
```
Comment (Tansito) [inline client/tests/core/test_nested_program_repository.py:11]:
```suggestion
    "http://localhost:80/v1/api/nested-programs/": {
```

File: client/tests/core/test_nested_program_repository.py [hunk 5]
```diff
@@ -102,7 +102,7 @@ def mocked_requests_get(**kwargs):
                 os.path.dirname(os.path.abspath(__file__)),
                 "..",
                 "resources",
-                "program.tar",
+                "nested_program.tar",
```
Comment (Tansito) [inline client/tests/core/test_nested_program_repository.py:105]: I think this `tar` is not being renamed.

**Review by Tansito — CHANGES_REQUESTED** [review]

**Comment by agarwal-aryan** [comment]
@Tansito 
Sir what is to be done now as I have done all the changes suggested by you?

**Comment by Tansito** [comment]
@agarwal-aryan probably it remains some minor changes with the linter, as the github action is specifying. You can just run `tox -elint` to check the errors and fix them.

**Comment by pacomf** [comment]
> @pacomf Do I have to change the name of the file also?

Yes @agarwal-aryan you should replace any program by nested_program included filenames if it is related. Check the test (like @Tansito mentioned, lint is still failing). You can check in your local running the command `tox -elint` , it will show all lint errors that you need to fix to pass tests :)

**Comment by agarwal-aryan** [comment]
@pacomf  
Now, `tox -elint` is showing ` lint: OK (0.09 seconds)
  congratulations :) (0.27 seconds)`

**Comment by pacomf** [comment]
But the code is still failing because you are importing files that didnt exist with that name (maybe you did Replace All Program -> NestedProgram in code, but references to filename needs to be nested_program)

You can run tests in local (not just lint) to check if it is something failing (`tox -epy39`, like the [contributing file](https://github.com/Qiskit-Extensions/quantum-serverless/blob/main/CONTRIBUTING.md#adding-tests) explains)

Basically, if you see the test failing (because name to import: https://github.com/Qiskit-Extensions/quantum-serverless/actions/runs/4510919410/jobs/7942829391?pr=347) for example, one of the errors is:

You are importing here (https://github.com/agarwal-aryan/quantum-serverless/blob/main/client/quantum_serverless/quantum_serverless.py#L38 -> `quantum_serverless.core.NestedProgram`) but the right file name in that path should be (https://github.com/agarwal-aryan/quantum-serverless/blob/main/client/quantum_serverless/core/nested_program.py -> `quantum_serverless.core.nested_program`)

Please, check all this kind of errors and ensure that tests are passing (you can double check in the section of Checks at the bottom of this PR)

You are doing a great job!

**Comment by agarwal-aryan** [comment]
@pacomf 
I have checked all the errors (acc. to best my knowledge) now please check the problems.

**Comment by pacomf** [comment]
@Tansito can you review it again now than the tests are passing?

File: client/quantum_serverless/quantum_serverless.py [hunk 6]
```diff
@@ -85,12 +85,12 @@ def job_client(self):
         """Job client for given provider."""
         return self._selected_provider.job_client()
 
-    def run_program(self, program: Program) -> Optional[Job]:
-        """Execute program as a async job
+    def run_nested_program(self, nested_program: NestedProgram) -> Optional[Job]:
```
Comment (IceKhan13) [inline client/quantum_serverless/quantum_serverless.py:88]: `run_nested_program` is kind long name :) 

Maybe just `run` or leave `run_program` as it was. @Tansito @pacomf  can you weight in on that.

**Comment by agarwal-aryan** [comment]
@IceKhan13 Sir
So, what should I have to do now?

**Comment by agarwal-aryan** [comment]
@Tansito Sir can you tell what should I have to do now?

**Comment by pacomf** [comment]
@agarwal-aryan we will perform the final review during this week! Thanks for your help! :)

File: client/quantum_serverless/core/job.py [hunk 7]
```diff
@@ -53,16 +53,16 @@
...
+        """Runs nested_program."""
         raise NotImplementedError
 
     def status(self, job_id: str):
         """Check status."""
         raise NotImplementedError
 
     def stop(self, job_id: str):
-        """Stops job/program."""
+        """Stops job/nested_program."""
```
Comment (Tansito) [inline client/quantum_serverless/core/job.py:65]:
```suggestion
        """Stops job/nested-program."""
```

File: client/quantum_serverless/quantum_serverless.py (hunk 6 above)
Comment (Tansito) [reply client/quantum_serverless/quantum_serverless.py:88]: @agarwal-aryan Let's use `run` only in this case as Iskandar was mentioning here 🙏

**Review by Tansito — CHANGES_REQUESTED** [review]
We are almost there @agarwal-aryan thank you 🙏 

It just remains a couple of comments to address to finish with the PR.

File: client/quantum_serverless/quantum_serverless.py (hunk 6 above)
Comment (Tansito) [inline client/quantum_serverless/quantum_serverless.py:88]: The idea @agarwal-aryan is change this line to:

```suggestion
    def run(self, nested_program: NestedProgram) -> Optional[Job]:
```

And review where is being used.

**Comment by psschwei** [comment]
image


that's weird... the lint checks disappeared...

**Comment by Tansito** [comment]
The service seems to be running again, rerunning github actions 🤞

**Review by Tansito — APPROVED** [review]
Now I see everything correct 😄  

Thank you @agarwal-aryan for your first contribution in `quantum-serverless` 🎉

**Review by pacomf — APPROVED** [review]

**Comment by agarwal-aryan** [comment]
Welcome @Tansito  and also Thank you 🙏🙂 to @pacomf , @Tansito , @IceKhan13 and @psschwei for reviewing my changes and answering my doubts and helping me to do my first contribution to `quantum-serverless`

**Comment by Tansito** [comment]
After a conversation yesterday with @pacomf we will need to take a look to the tutorials before merge this PR. I will open a PR and relate it with this one.

**Comment by Tansito** [comment]
Merging this PR in #390 to use it with the tutorials. Thank you again for your contribution @agarwal-aryan you will be able to see it in `main` soon 😄

**Comment by agarwal-aryan** [comment]
@Tansito Welcome

**Pull request merged**